### ISSUE TITLE
Misleading caller identity when `Which` fails

### ISSUE THREAD

**Issue by jnyrup** [fluentassertions/fluentassertions#1502]
### Description

When chaining into an assertion using `Which` the failure message can be confusing, when the inner assertion fails, as the determined caller identity is from the outer assertion.

### Complete minimal example reproducing the issue

```csharp
var languages = new[] { "C#" };

languages.Should().ContainSingle()
    .Which.Should().EndWith("script");
```

### Expected behavior:

Some nice message about "C#" not ending with "script".

### Actual behavior:

```
Expected languages to end with "script", but "C#" is too short.
```

### Versions

* latest version of develop: ee0aa439c542cbab2d83cdb41ea65849585114d0

**Comment by dennisdoomen** [comment]
Yeah, also noticed this while preparing the demo video.

**Comment by sdelarosbil** [comment]
I tried to look into this and it seems complex.

You would have to somehow have a way for the call identifier's parser have the ability to look ahead and see if we don't have a Which.Should call down the line which seems difficult to detect right on top of the identifier being dependant on the current assertion context.

The later seems doable, but I am really not sure how you'd tackle the former. Unless I am missing something, this doesn't seem easy to address.

**Comment by dennisdoomen** [comment]
Or `Which` should leave some clue to the caller identification code that it should not try to get the caller.

**Comment by dennisdoomen** [comment]
Assuming the following example:

> ```cs
> var languages = new[] { "C#" };
> 
> languages.Should().ContainSingle()
>     .Which.Should().EndWith("script");
> ```

I would expect a failing `EndWith` to throw the appropriate exception with

`Expected languages[0] to end with "script", but "C#" is too short.` 

or

`Expected languages.Single() to end with "script", but "C#" is too short.`

Either way, we need some way to understand what `ContainSingle().Which` will do. I can think of two solutions:

1. Make the `CallerIdentifier` smart enough so that it knows what the various methods like `ContainSingle` do and how that affects the identifier. 
2. Give the `ContainSingle` the means to pass along information on how to combine the variable `languages` extracted by `CallerIdentifier` with something specific like `[0]` or `Single()`. 

I don't like option 1, but option 2 most definitely will require a breaking change. I'm thinking of a way where each `Should` call will create an `AssertionScope` (either a new one, or one connected to the ambient scope) and pass that on to the assertion class and its assertion methods. In that way, a method like `ContainSingle` can update the scope with the relevant context, which is then passed on to the chained `Should`.

**Comment by comecme** [comment]
I suppose this would also fix #2407?

**Comment by dennisdoomen** [comment]
Will most likely be solvable after #2253
